infra/migrations/1711833444337_create-separate-balance-operations-tables.js

```diff
@@ -0,0 +1,298 @@
+      type: 'uuid',
+      default: pgm.func('gen_random_uuid()'),
+      notNull: true,
+      primaryKey: true,
+    },
+
+    sequence: {
+      type: 'serial',
+      notNull: true,
+    },
+
+    user_id: {
+      type: 'uuid',
+      notNull: true,
+    },
+
+    amount: {
+      type: 'integer',
+      notNull: true,
+    },
+
+    originator_type: {
+      type: 'text',
+      notNull: true,
+    },
+
+    originator_id: {
+      type: 'uuid',
+      notNull: true,
+    },
+
+    created_at: {
+      type: 'timestamp with time zone',
+      notNull: true,
+      default: pgm.func("(now() at time zone 'utc')"),
+    },
+  });
+
+  pgm.createTable('user_tabcash_operations', {
+    id: {
+      type: 'uuid',
+      default: pgm.func('gen_random_uuid()'),
+      notNull: true,
+      primaryKey: true,
+    },
+
+    sequence: {
+      type: 'serial',
+      notNull: true,
+    },
+
+    user_id: {
+      type: 'uuid',
+      notNull: true,
+    },
+
+    amount: {
+      type: 'integer',
+      notNull: true,
+    },
+
+    originator_type: {
+      type: 'text',
+      notNull: true,
+    },
+
+    originator_id: {
+      type: 'uuid',
+      notNull: true,
+    },
+
+    created_at: {
+      type: 'timestamp with time zone',
+      notNull: true,
+      default: pgm.func("(now() at time zone 'utc')"),
+    },
+  });
+
+  pgm.createFunction(
+    'get_content_current_tabcoins',
+    [
+      {
+        name: 'content_id_input',
+        mode: 'IN',
+        type: 'uuid',
+      },
+    ],
+    {
+      returns: 'integer',
+      language: 'plpgsql',
+      replace: true,
+    },
+    `
+    DECLARE
+      total_tabcoins integer;
+    BEGIN
+      SELECT COALESCE(SUM(amount), 0)
+      INTO total_tabcoins
+      FROM content_tabcoin_operations
+      WHERE
+        content_id = content_id_input;
+
+      RETURN total_tabcoins;
+    END;
+  `,
+  );
+
+  pgm.createFunction(
+    'get_user_current_tabcoins',
+    [
+      {
+        name: 'user_id_input',
+        mode: 'IN',
+        type: 'uuid',
+      },
+    ],
+    {
+      returns: 'integer',
+      language: 'plpgsql',
+      replace: true,
+    },
+    `
+    DECLARE
+      total_tabcoins integer;
+    BEGIN
+      SELECT COALESCE(SUM(amount), 0)
+      INTO total_tabcoins
+      FROM user_tabcoin_operations
+      WHERE
+        user_id = user_id_input;
+
+      RETURN total_tabcoins;
+    END;
+  `,
+  );
+
+  pgm.createFunction(
+    'get_user_current_tabcash',
+    [
+      {
+        name: 'user_id_input',
+        mode: 'IN',
+        type: 'uuid',
+      },
+    ],
+    {
+      returns: 'integer',
+      language: 'plpgsql',
+      replace: true,
+    },
+    `
+    DECLARE
+    total_tabcash integer;
+    BEGIN
+      SELECT COALESCE(SUM(amount), 0)
+      INTO total_tabcash
+      FROM user_tabcash_operations
+      WHERE
+        user_id = user_id_input;
+
+      RETURN total_tabcash;
+    END;
+  `,
+  );
+
+  pgm.createFunction(
+    'get_content_tabcoins_credit_debit',
```

Comment:
```suggestion
    'get_content_balance_credit_debit',
```

Usando o mesmo nome para a função, que já era específica para conteúdos, e mantendo o retorno como `total_balance`,  evitamos todas as modificações em `models/content` e `queries/rankingQueries`.

Enquanto conteúdos só tiverem um tipo de moeda, acho que `content_balance` permanece adequado.

Reply:
Realizar essa alteração modificando a função existente fará os testes quebrarem no PR da migração. De todo modo, vou subir as alterações para você revisar.

É viável realizar o merge com tudo em um único PR, considerando que a variável de ambiente pode ser definida para desabilitar as rotas (https://github.com/filipedeschamps/tabnews.com.br/pull/1662).

Reply:
> Realizar essa alteração modificando a função existente fará os testes quebrarem no PR da migração. De todo modo, vou subir as alterações para você revisar.

Certo, acho que é porque as mudanças nas funções do Postgres precisam ser feitas em conjunto com o restante das modificações, ou seja, nesse PR aqui, e não no PR de criação das novas tabelas.

Podemos mesclar esse aqui ainda em "modo leitura" até rodarmos as migrations desse aqui também, então fazemos o redeploy voltando ao "modo normal".

Além disso, pode ser interessante criar uma nova migration para já fazer a cópia dos dados no mesmo PR da criação da novas tabelas. Deixando para o processo manual, apenas as queries de validações se tudo deu certo. O que acha?

> É viável realizar o merge com tudo em um único PR, considerando que a variável de ambiente pode ser definida para desabilitar as rotas (#1662).

Tudo no mesmo PR não é possível porque entramos em um paradoxo. O build resultará em erro, pois não foram executadas as migrations, e as migrations não poderão ser executadas, porque o erro no build impedirá o deploy.

Isso porque, em homologação e produção, adotamos o processo de rodar as migrations sempre de forma manual. Existe outra abordagem de CI que roda as migrations no build, mas isso exigiria uma mudança na forma como testamos em homologação, exigindo um banco diferente (ou uma branch do banco, em serviços que permitem isso) para cada deploy em homologação.

Reply:
> Podemos mesclar esse aqui ainda em "modo leitura" até rodarmos as migrations desse aqui também, então fazemos o redeploy voltando ao "modo normal".

Sim, pensei nisso.

> Além disso, pode ser interessante criar uma nova migration para já fazer a cópia dos dados no mesmo PR da criação da novas tabelas. Deixando para o processo manual, apenas as queries de validações se tudo deu certo. O que acha?

Você está falando sobre as queries mencionadas em https://github.com/filipedeschamps/tabnews.com.br/pull/1656#issuecomment-2028481797, certo? Posso fazer isso. Crio um arquivo de migration novo apenas para isso, neste mesmo PR?

Outra pergunta, não tenho certeza como as queries irão se comportar se fizer tudo de uma vez ("sem `WHERE`"), principalmente considerando o tempo limite de execução da lambda. Faz sentido executar em blocos? Por exemplo, da `sequence = 1` em `balance_operations` até a `sequence = 100000`, e depois da `100001` até `200000` etc. (não sei se o ideal é de 100 mil em 100 mil).

> Tudo no mesmo PR não é possível porque entramos em um paradoxo. O build resultará em erro, pois não foram executadas as migrations, e as migrations não poderão ser executadas, porque o erro no build impedirá o deploy.

Não entendi. Quando eu disse "tudo em um único PR" quis dizer para "esquecermos" o PR #1656 (que está quebrando os testes), não sei se foi nesse ponto que me expressei mal. O ambiente de homologação está no ar, apenas sem funcionar algumas coisas (como o login) porque as migrações não foram executadas. Então o processo que imaginei era:

1. Mesclar os PRs #1663 e #1662.
2. Definir a variável de ambiente `UNDER_MAINTENANCE` para bloquear os endpoints `POST /api/v1/contents/*` e `GET /api/v1/user`.
3. Mesclar este PR.
4. Executar as migrações (de tabelas e dados).
5. Remover a variável de ambiente `UNDER_MAINTENANCE`.

Reply:
> > Além disso, pode ser interessante criar uma nova migration para já fazer a cópia dos dados no mesmo PR da criação da novas tabelas. Deixando para o processo manual, apenas as queries de validações se tudo deu certo. O que acha?
> 
> Você está falando sobre as queries mencionadas em [#1656 (comment)](https://github.com/filipedeschamps/tabnews.com.br/pull/1656#issuecomment-2028481797), certo?

Falava dessas mesmo!

> Posso fazer isso. Crio um arquivo de migration novo apenas para isso, neste mesmo PR?

Se der certo juntar tudo no mesmo PR, então poderia criar nesse aqui, mas eu falava sobre incluir no outro PR. Só que, pelo ponto que você comentou a seguir, é melhor deixar essa ideia de lado.

> Outra pergunta, não tenho certeza como as queries irão se comportar se fizer tudo de uma vez ("sem `WHERE`"), principalmente considerando o tempo limite de execução da lambda. Faz sentido executar em blocos? Por exemplo, da `sequence = 1` em `balance_operations` até a `sequence = 100000`, e depois da `100001` até `200000` etc. (não sei se o ideal é de 100 mil em 100 mil).

Acho que não adiantaria executar em blocos. Não tenho certeza, pois depende de como funciona por baixo a `node-pg-migrate`. Já são mais de 600 mil linhas em `balance_operations`, então deve ser melhor fazer a cópia manualmente mesmo. Nesse ponto só falta criar algumas consultas para checar se a cópia deu certo, né?

> > Tudo no mesmo PR não é possível porque entramos em um paradoxo. O build resultará em erro, pois não foram executadas as migrations, e as migrations não poderão ser executadas, porque o erro no build impedirá o deploy.
> 
> Não entendi. Quando eu disse "tudo em um único PR" quis dizer para "esquecermos" o PR #1656 (que está quebrando os testes), não sei se foi nesse ponto que me expressei mal. O ambiente de homologação está no ar, apenas sem funcionar algumas coisas (como o login) porque as migrações não foram executadas. Então o processo que imaginei era:
> 
> 1. Mesclar os PRs [feat(error message text): add `createErrorMessage` function #1663](https://github.com/filipedeschamps/tabnews.com.br/pull/1663) e [Permitir desabilitar rotas temporariamente pela variável de ambiente `UNDER_MAINTENANCE` #1662](https://github.com/filipedeschamps/tabnews.com.br/pull/1662).
> 2. Definir a variável de ambiente `UNDER_MAINTENANCE` para bloquear os endpoints `POST /api/v1/contents/*` e `GET /api/v1/user`.
> 3. Mesclar este PR.
> 4. Executar as migrações (de tabelas e dados).
> 5. Remover a variável de ambiente `UNDER_MAINTENANCE`.

Não tinha reparado que esse último commit permitiu o deploy em homologação. Como os anteriores estavam dando erro no build, então não daria pra fazer em um único PR.

Dessa forma, e com o "modo leitura", parece que podemos sim pular o PR #1656 👍
 
Daí não faz a cópia das tabelas pelas migrations, mas sim manualmente (ideia original), e depois fazemos as verificações.

Só então adequar a `UNDER_MAINTENANCE` (não pode ser removida, pois valerá a do `.env`), e fazer o redeploy.

Reply:
> então deve ser melhor fazer a cópia manualmente mesmo

Ok.

> Nesse ponto só falta criar algumas consultas para checar se a cópia deu certo, né?

Tem algo em mente para isso? O que eu fazia enquanto testava manualmente (com bem menos registros) era comparar os `id`s  "no olho" e somar as linhas de cada tabela para ver se dava a mesma quantidade do `balance_operations`.

Vou ver o que consigo elaborar para isso, e aí crio um comentário separado (neste PR) com as queries para copiar os registros e as queries para validar que deu certo.

> Só então adequar a UNDER_MAINTENANCE (não pode ser removida, pois valerá a do .env), e fazer o redeploy.

:+1: